File: src/main/java/com/googongill/aditory/controller/dto/user/SignupRequest.java
package com.googongill.aditory.controller.dto.user;

import com.googongill.aditory.domain.User;
import com.googongill.aditory.domain.enums.Role;
import com.googongill.aditory.domain.enums.SocialType;
import jakarta.validation.constraints.NotBlank;
import lombok.Builder;
import lombok.Data;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;

import java.util.List;

@Data
@Builder
public class SignupRequest {
    @NotBlank
    private String username;
    @NotBlank
    private String password;
    @NotBlank
    private String nickname;
    private String contact;
    private List<String> userCategories;

    public User toEntity() {
        BCryptPasswordEncoder bCryptPasswordEncoder = new BCryptPasswordEncoder();
        String encodedPassword = bCryptPasswordEncoder.encode(this.password);
        return new User(this.username, encodedPassword, Role.ROLE_USER, SocialType.LOCAL, this.nickname, this.contact);
    }
}

File: src/main/java/com/googongill/aditory/controller/dto/user/request/SignupRequest.kt
package com.googongill.aditory.controller.dto.user.request

import com.googongill.aditory.domain.User
import com.googongill.aditory.domain.enums.Role
import com.googongill.aditory.domain.enums.SocialType
import jakarta.validation.constraints.NotBlank
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder

data class SignupRequest(
    @NotBlank
    val username: String,
    @NotBlank
    val password: String,
    @NotBlank
    val nickname: String,
    val contact: String? = null,
    val userCategories: List<String>
) {

    fun toEntity(): User {
        val bCryptPasswordEncoder = BCryptPasswordEncoder()
        val encodedPassword = bCryptPasswordEncoder.encode(password)
        return User(username, encodedPassword, Role.ROLE_USER, SocialType.LOCAL, nickname, contact)
    }

}
